File: android/src/ti/airship/TiAutopilot.java
/* Copyright Airship and Contributors */

package ti.airship;

import android.content.Context;

import androidx.annotation.NonNull;

import com.urbanairship.AirshipConfigOptions;
import com.urbanairship.Autopilot;
import com.urbanairship.UAirship;
import com.urbanairship.analytics.Analytics;
import com.urbanairship.app.GlobalActivityMonitor;
import com.urbanairship.json.JsonMap;

import ti.airship.utils.ConfigUtils;
import ti.airship.utils.PluginLogger;
import ti.airship.utils.PluginStore;

public class TiAutopilot extends Autopilot {

    private static final String VERSION = "10.0.0";

    private AirshipConfigOptions configOptions;

    @Override
    public void onAirshipReady(@NonNull UAirship airship) {
        PluginLogger.setLogLevel(airship.getAirshipConfigOptions().logLevel);
        airship.getAnalytics().registerSDKExtension(Analytics.EXTENSION_TITANIUM, VERSION);

        TiAirshipListener listener = TiAirshipListener.newListener(GlobalActivityMonitor.shared(UAirship.getApplicationContext()));
        airship.getChannel().addChannelListener(listener);
        airship.getPushManager().addPushListener(listener);
        airship.getPushManager().setNotificationListener(listener);
    }

    @Override
    public boolean isReady(@NonNull Context context) {
        configOptions = loadConfig(context);
        if (configOptions == null) {
            return false;
        }

        try {
            configOptions.validate();
            return true;
        } catch (Exception e) {
            PluginLogger.error(e, "Unable to takeOff");
            return false;
        }
    }

    private AirshipConfigOptions loadConfig(@NonNull Context context) {
        JsonMap config = PluginStore.shared(context).getAirshipConfig();
        if (config == null) {
            return null;
        }
        return ConfigUtils.parseConfig(context, config);
    }

    @Override
    public AirshipConfigOptions createAirshipConfigOptions(@NonNull Context context) {
        return configOptions;
    }
}

File: android/src/ti/airship/TiAutopilot.kt
/* Copyright Airship and Contributors */
package ti.airship

import android.content.Context
import androidx.annotation.Keep
import com.urbanairship.AirshipConfigOptions
import com.urbanairship.Autopilot
import com.urbanairship.UAirship
import com.urbanairship.analytics.Analytics
import com.urbanairship.messagecenter.MessageCenter
import com.urbanairship.preferencecenter.PreferenceCenter
import ti.airship.utils.ConfigParser
import ti.airship.utils.PluginLogger
import ti.airship.utils.PluginStore

@Keep
class TiAutopilot : Autopilot() {

    companion object {
        private const val VERSION = "9.2.0"
    }

    private var configOptions: AirshipConfigOptions? = null

    override fun onAirshipReady(airship: UAirship) {
        PluginLogger.setLogLevel(airship.airshipConfigOptions.logLevel)
        airship.analytics.registerSDKExtension(Analytics.EXTENSION_TITANIUM, VERSION)

        val listener = TiAirshipListener(UAirship.getApplicationContext())
        airship.channel.addChannelListener(listener)
        airship.pushManager.addPushListener(listener)
        airship.pushManager.notificationListener = listener
        airship.deepLinkListener = listener
        MessageCenter.shared().inbox.addListener(listener)
        PreferenceCenter.shared().openListener = listener
    }

    override fun isReady(context: Context): Boolean {
        configOptions = loadConfig(context)
        return if (configOptions == null) {
            false
        } else try {
            configOptions!!.validate()
            true
        } catch (e: Exception) {
            PluginLogger.error(e, "Unable to takeOff")
            false
        }
    }

    private fun loadConfig(context: Context): AirshipConfigOptions? {
        val config = PluginStore.shared(context).airshipConfig ?: return null
        return ConfigParser.parse(context, config)
    }

    override fun createAirshipConfigOptions(context: Context): AirshipConfigOptions? {
        return configOptions
    }
}